File: app/src/main/java/com/yfbx/widgets/activity/BaseActivity.java
package com.yfbx.widgets.activity;

import android.content.pm.ActivityInfo;
import android.os.Bundle;
import android.support.annotation.LayoutRes;
import android.support.v7.app.AppCompatActivity;
import android.view.WindowManager;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

import butterknife.ButterKnife;
import butterknife.Unbinder;

/**
 * Author:Edward
 * Date:2018/3/15
 * Description:
 */
public abstract class BaseActivity extends AppCompatActivity {


    //记录所有活动的Activity
    private final List<BaseActivity> mActivities = new LinkedList<>();
    private Unbinder binder;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        setLandscape(false);
        super.onCreate(savedInstanceState);
        mActivities.add(this);
        setContentView(attachLayout());
        binder = ButterKnife.bind(this);
    }

    /**
     * 布局ID
     */
    @LayoutRes
    public abstract int attachLayout();


    /**
     * 全屏
     */
    protected void setFullScreen() {
        int flagFullscreen = WindowManager.LayoutParams.FLAG_FULLSCREEN;
        getWindow().setFlags(flagFullscreen, flagFullscreen);
    }

    /**
     * 屏幕常亮
     */
    protected void keepScreenOn() {
        getWindow().addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);
    }

    /**
     * 固定横竖屏
     */
    protected void setLandscape(boolean isLandscape) {
        int orientation = isLandscape ? ActivityInfo.SCREEN_ORIENTATION_LANDSCAPE : ActivityInfo.SCREEN_ORIENTATION_PORTRAIT;
        setRequestedOrientation(orientation);
    }


    /**
     * 关闭所有Activity
     */
    public void finishAll() {
        List<BaseActivity> copy;
        synchronized (mActivities) {
            copy = new ArrayList<BaseActivity>(mActivities);
        }
        for (BaseActivity activity : copy) {
            activity.finish();
        }
    }

    /**
     * 退出应用
     */
    public void exitApp() {
        finishAll();
        android.os.Process.killProcess(android.os.Process.myPid());
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        mActivities.remove(this);
        if (binder != null) {
            binder.unbind();
        }
    }
}

File: app/src/main/java/com/yfbx/widgets/activity/BaseActivity.kt
package com.yfbx.widgets.activity

import android.content.pm.ActivityInfo
import android.os.Bundle
import android.os.PersistableBundle
import android.widget.FrameLayout
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.Fragment
import com.billy.android.swipe.SwipeConsumer
import com.yfbx.widgets.util.slideBack
import kotlinx.android.synthetic.main.layout_toolbar.*
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.MainScope
import kotlinx.coroutines.cancel

/**
 * Author: Edward
 * Date: 2019-11-08
 * Description:
 */
abstract class BaseActivity : AppCompatActivity(), CoroutineScope by MainScope() {

    private var oldFrag: Fragment? = null
    private var slideBack: SwipeConsumer? = null

    override fun onCreate(savedInstanceState: Bundle?, persistentState: PersistableBundle?) {
        requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_PORTRAIT
        super.onCreate(savedInstanceState, persistentState)
        slideBack = slideBack()
    }


    fun enableSlideBack(enable: Boolean) {
        if (enable) {
            slideBack?.enableLeft()
        } else {
            slideBack?.disableLeft()
        }
    }

    /**
     * Toolbar
     */
    protected fun setToolbar(title: String, back: Boolean = true) {
        setSupportActionBar(toolbar)
        supportActionBar?.setDisplayShowTitleEnabled(false)
        supportActionBar?.setDisplayHomeAsUpEnabled(back)
        toolbar.setNavigationOnClickListener { onBackPressed() }
        toolbarTxt.text = title
    }

    protected fun contentView(): FrameLayout {
        return findViewById(android.R.id.content)
    }

    /**
     * 切换Fragment
     */
    protected fun switchFragment(fragment: Fragment) {
        val transaction = supportFragmentManager.beginTransaction()
        if (!fragment.isAdded) {
            //transaction.add(R.id.content_view, fragment)
        }
        if (oldFrag != null) {
            transaction.hide(oldFrag!!)
        }
        transaction.show(fragment)
        oldFrag = fragment
        transaction.commit()
    }

    override fun onDestroy() {
        cancel()
        super.onDestroy()
    }
}